src/sparse/merkle_tree.rs

```diff
@@ -92,54 +92,63 @@ where
+        //    common prefix count (CPC).
+        // 2. Create a parent for the leaf node we traversed to and the leaf we are updating.
+        //    Invert the depth to get the parent's height.
+        // 3. Check the requested leaf's bit at the CPC to determine it's orientation relative
+        //    to the radix ancestor. If the requested leaf is on the left of the radix ancestor,
+        //    it is the parent's left child; otherwise, it is the parent's right child.
+
+        let ancestor_depth = {
             if actual_leaf_node.is_placeholder() {
-                self.depth()
+                // If the actual leaf node is a placeholder, we cannot infer any path instructions
+                // from its leaf key. Therefore, the only common ancestor is the root at depth 0.
+                0
             } else {
                 let actual_leaf_key = actual_leaf_node.leaf_key();
                 let requested_leaf_key = requested_leaf_node.leaf_key();
                 actual_leaf_key.common_prefix_count(requested_leaf_key)
             }
         };
+        let ancestor_height = self.max_height() - ancestor_depth;
 
-        if common_prefix_count != self.depth() {
-            let requested_leaf_key = requested_leaf_node.leaf_key();
-            if requested_leaf_key
-                .get_bit_at_index_from_msb(common_prefix_count)
+        // If the ancestor is not the root:
+        if ancestor_depth != 0 {
+            if requested_leaf_node
+                .leaf_key()
+                .get_bit_at_index_from_msb(ancestor_depth)
                 .unwrap()
                 == 1
             {
-                current_node = Node::create_node(&actual_leaf_node, &current_node);
+                current_node = Node::create_node(&actual_leaf_node, &requested_leaf_node);
             } else {
-                current_node = Node::create_node(&current_node, &actual_leaf_node);
+                current_node = Node::create_node(&requested_leaf_node, &actual_leaf_node);
             }
-            current_node.set_height((self.depth() - common_prefix_count) as u32);
+            current_node.set_height(ancestor_height as u32);
             self.storage
                 .insert(&current_node.hash(), current_node.as_buffer())?;
         }
 
-        let offset_side_nodes = self.depth() - side_nodes.len();
-        for i in 0..self.depth() {
+        let depth = std::cmp::max(side_nodes.len(), ancestor_depth);
+        let placeholders = depth - side_nodes.len();
+
+        for current_depth in 0..depth {
             let side_node = {
-                if i < offset_side_nodes {
-                    if common_prefix_count != self.depth()
-                        && common_prefix_count > self.depth() - 1 - i
-                    {
-                        Node::create_placeholder()
-                    } else {
-                        continue;
-                    }
```

Comment:
This assignment was needlessly complex. Deeper analysis of how these branches behave shows that we can reduce the iteration from 0 to 256 (`self.depth()`) to a more precise range, and remove the `continue` branch altogether. In the end, these loops yield the same result. 

The benefits of this are twofold:
- The iteration range is necessarily smaller on average
- The code communicates intent much more clearly, and is more maintainable